File: cowork-roadmap/src/test/java/com/cowork/roadmap/global/team/ProjectionSnapshotCompletionTest.java
package com.cowork.roadmap.global.team;

import static org.assertj.core.api.Assertions.assertThat;

import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.junit.jupiter.api.Test;

import tools.jackson.databind.json.JsonMapper;

class ProjectionSnapshotCompletionTest {

    @Test
    void validatesTheSharedPerPartitionSnapshotMarkerContract() {
        ConsumerRecord<String, String> record = new ConsumerRecord<>("team.member.event",
                1,
                7L,
                ProjectionSnapshotCompletion.KEY_PREFIX + "1",
                """
                        {"eventType":"PROJECTION_SNAPSHOT_COMPLETED","topic":"team.member.event","partition":1,
                         "snapshotId":"00000000-0000-0000-0000-000000000001",
                         "occurredAt":"2026-08-26T00:00:00Z","source":"cowork-team"}
                        """);

        assertThat(ProjectionSnapshotCompletion.violation(JsonMapper.builder().build(), record)).isNull();
    }

    @Test
    void mismatchedReservedMarkerNeverInitializesTheProjection() {
        ConsumerRecord<String, String> record = new ConsumerRecord<>("team.member.event",
                1,
                7L,
                ProjectionSnapshotCompletion.KEY_PREFIX + "0",
                "{}");

        assertThat(ProjectionSnapshotCompletion.violation(JsonMapper.builder().build(), record)).contains("partition");
    }

    @Test
    void markerFromAnotherProducerNeverInitializesTheProjection() {
        ConsumerRecord<String, String> record = new ConsumerRecord<>("team.member.event",
                1,
                7L,
                ProjectionSnapshotCompletion.KEY_PREFIX + "1",
                """
                        {"eventType":"PROJECTION_SNAPSHOT_COMPLETED","topic":"team.member.event","partition":1,
                         "snapshotId":"00000000-0000-0000-0000-000000000001",
                         "occurredAt":"2026-08-26T00:00:00Z","source":"cowork-deprecated"}
                        """);

        assertThat(ProjectionSnapshotCompletion.violation(JsonMapper.builder().build(), record))
                .contains("cowork-team");
    }
}

File: cowork-preference/src/test/kotlin/com/cowork/preference/messaging/ProjectionSnapshotCompletionTest.kt
package com.cowork.preference.messaging

import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class ProjectionSnapshotCompletionTest {
    private val payload =
        """{"eventType":"PROJECTION_SNAPSHOT_COMPLETED","topic":"team.member.event","partition":1,"snapshotId":"00000000-0000-0000-0000-000000000001","occurredAt":"2026-08-26T00:00:00Z","source":"cowork-team"}"""

    @Test
    fun `valid shared marker contract is accepted`() {
        assertNull(
            ProjectionSnapshotCompletion.violation(
                "${PreferenceEvents.SNAPSHOT_COMPLETED_KEY_PREFIX}1",
                payload,
                "team.member.event",
                1,
            ),
        )
    }

    @Test
    fun `reserved marker with wrong partition never completes initialization`() {
        val violation = ProjectionSnapshotCompletion.violation(
            "${PreferenceEvents.SNAPSHOT_COMPLETED_KEY_PREFIX}0",
            payload,
            "team.member.event",
            1,
        )

        assertTrue(requireNotNull(violation).contains("partition"))
    }

    @Test
    fun `marker from another producer never completes initialization`() {
        val violation = ProjectionSnapshotCompletion.violation(
            "${PreferenceEvents.SNAPSHOT_COMPLETED_KEY_PREFIX}1",
            payload.replace("cowork-team", "cowork-deprecated"),
            "team.member.event",
            1,
        )

        assertTrue(requireNotNull(violation).contains("cowork-team"))
    }
}
